### PR TITLE
modules: mcuboot: Allow multi-image to be used without BOOT_UPGRADE_ONLY

### PR DESCRIPTION
**Pull request by sigvartmh**
Using mult-image with external flash on nRF53 without `BOOT_UPGRADE_ONLY` will corrupt the network core when a revert is preformed as it will copy data from the simulated flash(RAM) which could be garbage data. It will also copy garbage data over the image stored in external flash when preforming the swap.

However there is a use-case where you would want revertable application core and you are not too concerned about what happens to the network core. This requires you have a method of loading new network core data into external flash from your application core application.

Ref. NCSDK-16932

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

File: modules/mcuboot/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,26 @@ if (CONFIG_MCUBOOT)
   if (CONFIG_NRF53_MULTI_IMAGE_UPDATE)
     set_shared(IMAGE ${IMAGE_NAME} PROPERTY NRF53_MULTI_IMAGE_UPDATE
       ${CONFIG_NRF53_MULTI_IMAGE_UPDATE})
+  if (NOT DEFINED CONFIG_BOOT_UPGRADE_ONLY)
+    # Added color defines due to this coming from a childe image and being out
+    # putted as raw unformatted text to increase visibility.
+    string(ASCII 27 Esc)
+    set(ColorReset "${Esc}[m")
+    set(BoldRed     "${Esc}[1;31m")
+    message(WARNING "${BoldRed}
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:15]: have you tested this formatting in windows ?

File: modules/mcuboot/CMakeLists.txt [hunk 2]
```diff
@@ -6,6 +6,26 @@ if (CONFIG_MCUBOOT)
...
+        ---------------------------------------------------------
+        --- WARNING: Using multi-image update on nRF53        ---
+        --- without BOOT_UPGRADE_ONLY corrupts the network    ---
+        --- upon image reverts. Only application core can be  ---
+        --- reverted.                                         ---
+        ---------------------------------------------------------
+	\n${ColorReset}"
+        )
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:23]: why is such a huge banner needed ?

Why not just a single `WARNING` line.
A single line of warning will still provide users with a stack trace, and should get their attention.

It's not like their device is being destroyed if they do this, they can still re-flash and one must expect our users to have examined any warnings carefully and tested their product before putting software into production.

A regular warning is sufficient here.

**Review by tejlmand — CHANGES_REQUESTED** [review]
The commit message says:
> This requires you have a method of loading new network core data
> into external flash from your application core application.

so how does the user informs the build system that such a method is in place ?
Users should be able to inform the build system that they know what they are doing, and thereby silence the `warning`.

Or simply have another prompt-less setting that users app can forcefully select:
```
depends on BOOT_UPGRADE_ONLY || I_KNOW_WHAT_IM_DOING_FLAG
```

then user's app `Kconfig` can do:
```
# <app>/Kconfig

config MY_APP
    default y
    select I_KNOW_WHAT_IM_DOING_FLAG
```

**Review by nvlsianpu — COMMENTED** [review]
Is this tested? Does network core upgrade works?

File: modules/mcuboot/CMakeLists.txt (hunk 1 above)
Comment (sigvartmh) [reply modules/mcuboot/CMakeLists.txt:15]: Nope. Could potentially not work?

**Comment by sigvartmh** [comment]
I've tested it and network update works with this patch.

File: modules/mcuboot/CMakeLists.txt (hunk 2 above)
Comment (sigvartmh) [reply modules/mcuboot/CMakeLists.txt:23]: > The commit message says:
> 
> > This requires you have a method of loading new network core data
> > into external flash from your application core application.
> 
> so how does the user informs the build system that such a method is in place ? Users should be able to inform the build system that they know what they are doing, and thereby silence the `warning`.

The build system currently does not, and have no way of disabling the warning good point.

> 
> Or simply have another prompt-less setting that users app can forcefully select:
> 
> ```
> depends on BOOT_UPGRADE_ONLY || I_KNOW_WHAT_IM_DOING_FLAG
> ```
> 
> then user's app `Kconfig` can do:
> 
> ```
> # <app>/Kconfig
> 
> config MY_APP
>     default y
>     select I_KNOW_WHAT_IM_DOING_FLAG
> ```

I think having some promtless setting is the way to go and have it documented in `.rts` instead. Thank you for the ide @tejlmand I'll refactor the PR.

**Comment by sigvartmh** [comment]
@FrancescoSer I didn't really find a good place to put a documentation of this feature. I was thinking the nrf5340 user-guide but it didn't really seem to fit there. I kind of want it linked with MCUBoot documentation to be honest.

**Comment by FrancescoSer** [comment]
> @FrancescoSer I didn't really find a good place to put a documentation of this feature. I was thinking the nrf5340 user-guide but it didn't really seem to fit there. I kind of want it linked with MCUBoot documentation to be honest.

As discussed offline, let's add the content in ug_5340.rst

File: modules/mcuboot/boot/zephyr/Kconfig [hunk 3]
```diff
@@ -108,7 +108,7 @@ config NRF53_MULTI_IMAGE_UPDATE
 	depends on FLASH_SIMULATOR_DOUBLE_WRITES
 	depends on !FLASH_SIMULATOR_STATS
 	depends on PM_EXTERNAL_FLASH_MCUBOOT_SECONDARY
-	depends on BOOT_UPGRADE_ONLY
+	depends on BOOT_UPGRADE_ONLY || USE_NRF53_MULIT_IMAGE_WITHOUT_UPGRADE_ONLY
```
Comment (tejlmand) [inline modules/mcuboot/boot/zephyr/Kconfig:111]: Did you test this ?

I'm surprised if it worked, cause there is a spelling mistake:
```suggestion
	depends on BOOT_UPGRADE_ONLY || USE_NRF53_MULTI_IMAGE_WITHOUT_UPGRADE_ONLY
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
How was this hidden setting tested to be working ?
The dependency had a spelling mistake so how could you ensure it was overwritten correctly.

File: modules/mcuboot/boot/zephyr/Kconfig (hunk 3 above)
Comment (sigvartmh) [reply modules/mcuboot/boot/zephyr/Kconfig:111]: Tested it but had the spelling mistake in both my custom Kconfig ^^

Comment (tejlmand) [reply modules/mcuboot/boot/zephyr/Kconfig:111]: so you want to keep the spelling mistake ?

Comment (sigvartmh) [reply modules/mcuboot/boot/zephyr/Kconfig:111]: Woops not I just haven't updated the PR :) Will update it with other changes :+1:

**Comment by sigvartmh** [comment]
@FrancescoSer Tried to add a note about the option but I'm not sure if it's enough information.

File: doc/nrf/ug_nrf5340.rst [hunk 4]
```diff
@@ -552,6 +552,12 @@ To enable the simultaneous update of multiple images in the MCUboot, set the fol
 * :kconfig:option:`CONFIG_PCD_APP` - Enable commands exchange with the network core.
 * :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` - Enable support for multiple update partitions by setting this option to ``2``.
 
+.. note::
+
+   The application core can be reverted however this will brick the network core upon revert as it will fill the network core with whatever is currently in the RAM that pcd uses.
+   To enable this the Kconfig ``USE_NRF53_MULTI_IMAGE_WITHOUT_UPGRADE_ONLY`` needs to be defined in the project level KConfig.
```
Comment (FrancescoSer) [inline doc/nrf/ug_nrf5340.rst:558]:
```suggestion
   To enable this, define the ``USE_NRF53_MULTI_IMAGE_WITHOUT_UPGRADE_ONLY`` kconfig option in the project-level KConfig file.
```

File: doc/nrf/ug_nrf5340.rst [hunk 5]
```diff
@@ -552,6 +552,12 @@ To enable the simultaneous update of multiple images in the MCUboot, set the fol
...
+   When this is option is defined it can be enabled setting :kconfig:option`CONFIG_USE_NRF53_MULTI_IMAGE_WITHOUT_UPGRADE_ONLY`.
```
Comment (FrancescoSer) [inline doc/nrf/ug_nrf5340.rst:559]:
```suggestion
   When this is option is defined, you can enable it by setting :kconfig:option`CONFIG_USE_NRF53_MULTI_IMAGE_WITHOUT_UPGRADE_ONLY`.
```

File: doc/nrf/ug_nrf5340.rst [hunk 6]
```diff
@@ -552,6 +552,12 @@ To enable the simultaneous update of multiple images in the MCUboot, set the fol
 * :kconfig:option:`CONFIG_PCD_APP` - Enable commands exchange with the network core.
 * :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` - Enable support for multiple update partitions by setting this option to ``2``.
 
+.. note::
+
+   The application core can be reverted however this will brick the network core upon revert as it will fill the network core with whatever is currently in the RAM that pcd uses.
```
Comment (FrancescoSer) [inline doc/nrf/ug_nrf5340.rst:557]:
```suggestion
   The application core can be reverted, but doing so bricks the network core upon revertion, as the reversion process fills the network core with the content currently in the RAM that pcd uses.
```

**Review by FrancescoSer — CHANGES_REQUESTED** [review]

File: doc/nrf/ug_nrf5340.rst [hunk 7]
```diff
@@ -574,6 +580,7 @@ Additionally, the memory partitions must be defined and include:
 Samples and applications built for Thingy:53 enable simultaneous update of multiple images by default.
 To learn more about Thingy:53, see :ref:`ug_thingy53`.
 
+
```
Comment (FrancescoSer) [inline doc/nrf/ug_nrf5340.rst:583]:
```suggestion
```

Comment (FrancescoSer) [reply doc/nrf/ug_nrf5340.rst:583]: This looks like a typo

**Review by FrancescoSer — APPROVED** [review]
Approving but please implement the small suggestion

**Review by tejlmand — APPROVED** [review]
lgtm

**Pull request merged**